cli/cmd/verify.go

```diff
@@ -96,7 +97,13 @@ func runVerify(cmd *cobra.Command, _ []string) error {
 	}
 	log.Debug("Got response")
 
-	fmt.Fprintln(cmd.OutOrStdout(), "✔️ Successfully verified coordinator")
+	currentManifest := resp.Manifests[len(resp.Manifests)-1]
+	// Local manifest will have a newline if the user has edited the file.
```

Comment:
Why isn't that the case for the manifest we set?

Reply:
This could (*Presumably*) be an editor-specific problem. I know that VSCode does append a newline to files by default upon editing. For the default manifest, if not edited, this should not be the case.  
I think we might want to strip all whitespace here for a canonical comparison.

Reply:
I think we should be verifying the manifest byte-for-byte here, and rather fix the JSON encoding in `generate`.

Reply:
The manifest is encoded in the GetManifest response, so it does not matter how the encoding is in the `generate` step. If we want to compare the manifest byte-for-byte, we need to add or remove a newline somewhere in the verify step or the GetManifest encoding, or convert to a canonical JSON encoding when comparing the manifests (e.g. `json.Compact`)

Reply:
Ah, I see. This should not be the case anymore since #613 - can you try with that base?

Reply:
Still an issue, since we are comparing the marshaled manifest to the bytes read from the file. I also noticed that the `SeedShareOwnerPubKeys` field is null after `generate` but then missing in the obtained history from `verify`...

Reply:
> Still an issue, since we are comparing the marshaled manifest to the bytes read from the file.

There should be no marshalling anymore - the manifest should make the full round-trip as `[]byte`, and if it doesn't that's a bug I need to fix.

> I also noticed that the SeedShareOwnerPubKeys field is null after generate but then missing in the obtained history from verify...

This code path is not complete yet, I'm afraid.

Reply:
No marshalling except for the initial marshalling in `generate`, so yeah, we can fix the encoding here.

> This code path is not complete yet, I'm afraid.

Forgot to rebuild, now it works.

But is it a good idea to compare byte-for-byte? The verification then depends on, e.g., the user's JSON indentation.

Reply:
canonicalization might be a good idea, feel free to create a ticket. I don't think the policies have a canonical form.

Reply:
The `SetManifest` method [accepts manifest and policies as `[]byte`](https://github.com/edgelesssys/contrast/blob/afdc62e96a220d43f807fcfb28013904057abf9d/internal/userapi/userapi.proto#L13-L14) and the `GetManifests` method [returns them as `[]byte`](https://github.com/edgelesssys/contrast/blob/afdc62e96a220d43f807fcfb28013904057abf9d/internal/userapi/userapi.proto#L39-L40). We don't need to care about the user's JSON indentation if we ensure that the manifest can round trip without being recoded. 

> I don't think the policies have a canonical form.

They don't, but the policy hash requires them being byte-for-byte identical.

Reply:
> We don't need to care about the user's JSON indentation if we ensure that the manifest can round trip without being recoded

Changes could occur if the manifest is handed out-of-band from the workload owner to the relying party and then used on verify.

> They don't, but the policy hash requires them being byte-for-byte identical.

Canonicalization would be applied (if it would exist) before hashing, so that wouldn't be a problem.